### DISCUSSION TITLE
Using Quarkus Extensions in Keycloak.X

### DISCUSSION THREAD

**Discussion by fbuetler** [keycloak/keycloak#13504] [Archive]
Hi,

We would like to use an additional Quarkus extensions ([quarkus-logging-logback](https://github.com/quarkiverse/quarkus-logging-logback)) in our Keycloak.X distribution (version 18.0.2).

We tried the approach as disccued here https://github.com/keycloak/keycloak/discussions/10884#discussioncomment-2468811 by copying the `quarkus-logging-logback`, `quarkus-logging-logback-deployment` and other dependencies into the `providers` folder with a subsequent `kc.sh build`, but without any luck.

We also found this issue here [KEYCLOAK-18475](https://issues.redhat.com/browse/KEYCLOAK-18475), that implicitly states custom dependencies are supported but Quarkus extensions are not.

Unfortunately, the documentation only talks about [dependencies for providers](https://www.keycloak.org/server/configuration-provider#_using_third_party_dependencies).

Any help is appreciated!

**Comment by blevine** [comment]
This is old, but wondering whether there was any progress on this. I'm trying to integrate quarkus-websockets-next into Keycloak 26.0.1 without any luck.

**Comment by bernhof** [comment]
Also very interested in this. Our use case is support for logging in ECS (Elastic Common Schema), which comes with [quarkus-logging-json](https://quarkus.io/guides/logging#quarkus-logging-json_quarkus-log-file-json-log-format) since Keycloak does not easily allow us to customize the json output.

Relates to #33959 #10414 #8870

**Reply by mabartos** [reply]
You can leverage this [keycloak-quarkus-extensions](https://github.com/mabartos/keycloak-quarkus-extensions) tool to enhance your Keycloak deployment and add your Quarkus/Quarkiverse extension.

It is possible to support ECS in Keycloak by enhancing your Keycloak distribution :champagne: :champagne: 

Please check the [Logging JSON ECS](https://github.com/mabartos/keycloak-quarkus-extensions/blob/main/examples/logging-json-ecs.md) guide.

**Comment by mabartos** [comment]
Hey guys, I've started a project [keycloak-quarkus-extensions](https://github.com/mabartos/keycloak-quarkus-extensions) to be able to use whatever Quarkus/Quarkiverse extension in your Keycloak deployment. 

How to build an extended Keycloak distribution:

```
git clone git@github.com:mabartos/keycloak-quarkus-extensions.git
cd keycloak-quarkus-extensions/

./kc-extensions.sh add quarkus-logging-json

./kc-extensions.sh build
./kc-extensions.sh start-dev
```
You need to edit the `application.properties` with Quarkus properties for the logging.

After the build phase, `keycloak-extended-26.0.5.zip` is generated and contains an extended distribution.

Note: the project is still in development and will be improved during/after Xmas :)) Keep an eye on it if you're interested in providing such functionality.

**Reply by bernhof** [reply]
Cool that you've found a way. I'd need some guidance, though, on how to use this as part of a docker image build. Currently, we're creating a custom image based on quay.io/keycloak/keycloak, adding custom SPIs, healthcheck etc. in a Dockerfile consisting of "build" phase and finally starting Keycloak in optimized mode. Not entirely clear for me how to best use your tool in this situation. Since kc-extensions.sh commands (build/start-dev) seem to replace those of kc.sh, it seems to me that kc-extensions.sh should be included in the container/image build and commands run there rather than on the host machine. This would also mean that pom.xml isn't available (though this isn't a problem in our case)

This seems like a normal use case, and something that would be helpful to have documented specifically. :)

Look forward to following progress on this 👍

**Reply by mabartos** [reply]
Yes, I want to add the guidance for leveraging Dockerfile and have it as smooth as possible. But as I've mentioned, the complete solution with examples will be done after Christmas :)

**Reply by bernhof** [reply]
Do you already have an idea about how go about this? If I find the time, I'd gladly contribute som documentation if I can get it to work.

**Reply by mabartos** [reply]
You can leverage this [keycloak-quarkus-extensions](https://github.com/mabartos/keycloak-quarkus-extensions) tool to enhance your Keycloak deployment and add your Quarkus/Quarkiverse extension.

It is possible to support ECS in Keycloak by enhancing your Keycloak distribution :champagne: :champagne: 

Please check the [Logging JSON ECS](https://github.com/mabartos/keycloak-quarkus-extensions/blob/main/examples/logging-json-ecs.md) guide.
